<a href="https://colab.research.google.com/github/muinul-official/Data-Mining-G13/blob/hannah-feature-engineering/EDA_%26_Feature_Engineering.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# WIE3007 Data Mining Group Project
##EDA & Feature Engineering

* Name: Hannah bt Arham Rahimy
* Matrics Number : 22002434
* Role: Data Cleaning, Encoding, Feature Engineering, AI-based Feature Extraction



###1) Load Dataset

In [1]:
import pandas as pd
churn = pd.read_csv("ecommerce_churn_dataset.csv")
transactions = pd.read_csv("ecommerce_transactions.csv")
chats = pd.read_csv("ecommerce_chat_logs.csv")
reviews = pd.read_csv("ecommerce_reviews.csv")

print("churn:", churn.shape)
print("transactions:", transactions.shape)
print("chats:", chats.shape)
print("reviews:", reviews.shape)

churn.head()

churn: (1200, 30)
transactions: (6396, 8)
chats: (1200, 5)
reviews: (1200, 5)


customer_id  age  gender  location membership_tier  account_age_days  \
0  CUST000001   58    Male     Urban            Gold               480   
1  CUST000002   18    Male     Rural        Platinum               382   
2  CUST000003   73  Female     Rural          Silver                 9   
3  CUST000004   61   Other     Urban          Silver               184   
4  CUST000005   28   Other  Suburban          Bronze               336   

   recency  frequency  monetary_value  avg_order_value  ...  \
0        0          2           88.15            44.08  ...   
1        0          2          245.31           122.66  ...   
2        9          1           18.85            18.85  ...   
3        0          6          126.75            21.12  ...   
4        0          3           67.16            22.39  ...   

                                         review_text  review_rating  \
0                 Average purchase. No major issues.              3   
1                    It's fine. Nothing exceptional.              3   
2  The worst purchase I've made. Don't buy this p...              2   
3  Product doesn't work as advertised. Very frust...              2   
4  Wonderful product. Fast delivery. Highly recom...              4   

   review_sentiment_score  review_length  sentiment_score_avg  \
0                   0.520             34                0.533   
1                   0.594             31                0.499   
2                   0.055             53                0.055   
3                   0.110             53                0.099   
4                   0.714             51                0.796   

   text_engagement_score total_spent_per_day  order_frequency_per_month  \
0                  0.445            0.183264                   0.117647   
1                  0.440            0.640496                   0.145631   
2                  0.610            1.885000                   0.769231   
3                  0.635            0.685135                   0.841121   
4                  0.585            0.199288                   0.245902   

   churn  churn_probability  
0      0              0.330  
1      0              0.436  
2      0              0.489  
3      1              0.577  
4      0              0.158  

[5 rows x 30 columns]

###2) EDA

In [4]:
churn.columns

Index(['customer_id', 'age', 'gender', 'location', 'membership_tier',
       'account_age_days', 'recency', 'frequency', 'monetary_value',
       'avg_order_value', 'total_orders', 'days_since_last_login',
       'cart_abandonment_rate', 'email_open_rate', 'support_tickets',
       'frustration_index', 'chat_log', 'chat_sentiment_score', 'chat_length',
       'has_negative_keywords', 'review_text', 'review_rating',
       'review_sentiment_score', 'review_length', 'sentiment_score_avg',
       'text_engagement_score', 'total_spent_per_day',
       'order_frequency_per_month', 'churn', 'churn_probability'],
      dtype='object')

In [5]:
churn.head()

customer_id  age  gender  location membership_tier  account_age_days  \
0  CUST000001   58    Male     Urban            Gold               480   
1  CUST000002   18    Male     Rural        Platinum               382   
2  CUST000003   73  Female     Rural          Silver                 9   
3  CUST000004   61   Other     Urban          Silver               184   
4  CUST000005   28   Other  Suburban          Bronze               336   

   recency  frequency  monetary_value  avg_order_value  ...  \
0        0          2           88.15            44.08  ...   
1        0          2          245.31           122.66  ...   
2        9          1           18.85            18.85  ...   
3        0          6          126.75            21.12  ...   
4        0          3           67.16            22.39  ...   

                                         review_text  review_rating  \
0                 Average purchase. No major issues.              3   
1                    It's fine. Nothing exceptional.              3   
2  The worst purchase I've made. Don't buy this p...              2   
3  Product doesn't work as advertised. Very frust...              2   
4  Wonderful product. Fast delivery. Highly recom...              4   

   review_sentiment_score  review_length  sentiment_score_avg  \
0                   0.520             34                0.533   
1                   0.594             31                0.499   
2                   0.055             53                0.055   
3                   0.110             53                0.099   
4                   0.714             51                0.796   

   text_engagement_score total_spent_per_day  order_frequency_per_month  \
0                  0.445            0.183264                   0.117647   
1                  0.440            0.640496                   0.145631   
2                  0.610            1.885000                   0.769231   
3                  0.635            0.685135                   0.841121   
4                  0.585            0.199288                   0.245902   

   churn  churn_probability  
0      0              0.330  
1      0              0.436  
2      0              0.489  
3      1              0.577  
4      0              0.158  

[5 rows x 30 columns]

In [16]:
#target variable understanding
churn['churn'].value_counts()

churn
0    925
1    275
Name: count, dtype: int64

In [10]:
#check if customer_id exists in all datasets
transactions.columns

Index(['customer_id', 'transaction_date', 'amount', 'product_category',
       'quantity', 'discount_applied', 'payment_method', 'shipping_method'],
      dtype='object')

In [11]:
chats.columns

Index(['customer_id', 'chat_log', 'chat_sentiment_score', 'chat_length',
       'has_negative_keywords'],
      dtype='object')

In [12]:
reviews.columns

Index(['customer_id', 'review_text', 'review_rating', 'review_sentiment_score',
       'review_length'],
      dtype='object')

In [25]:
#check dataset shape & missing values
transactions.shape, transactions.isnull().sum()

((6396, 8),
 customer_id         0
 transaction_date    0
 amount              0
 product_category    0
 quantity            0
 dtype: int64)

In [26]:
chats.shape, chats.isnull().sum()

((1200, 5),
 customer_id              0
 chat_log                 0
 chat_sentiment_score     0
 chat_length              0
 has_negative_keywords    0
 dtype: int64)

In [27]:
reviews.shape, reviews.isnull().sum()

((1200, 5),
 customer_id               0
 review_text               0
 review_rating             0
 review_sentiment_score    0
 review_length             0
 dtype: int64)

#### EDA Summary

The main churn dataset contains 1,200 customer-level records with a binary churn target, where 925 customers are non-churned and 275 are churned, indicating moderate class imbalance.

Supporting datasets include transaction history (6396 records), customer support chat logs (1200), and product reviews (1200). Transaction data is recorded at transaction level, while chat and review datasets are already aggregated at customer level.

All datasets share a common `customer_id`, enabling integration across data sources. Initial inspection shows no missing values in key variables, indicating good data quality for feature engineering and modeling.


###3) Data Cleaning & Preprocessing

In [18]:
# correcting data types
transactions['amount'] = pd.to_numeric(transactions['amount'], errors='coerce')

In [20]:
# check duplicate
transactions.duplicated().sum()

np.int64(0)

In [22]:
chats.duplicated().sum()

np.int64(0)

In [23]:
reviews.duplicated().sum()

np.int64(0)

In [21]:
transactions.describe()

amount     quantity
count  6396.000000  6396.000000
mean     69.715044     3.020951
std     131.189101     1.417924
min       0.400000     1.000000
25%      14.812500     2.000000
50%      33.320000     3.000000
75%      74.935000     4.000000
max    3683.010000     5.000000

####Data Cleaning & Preprocessing Summary :

Basic preprocessing was applied, including data type correction, duplicate checks,
and sanity inspection of numerical values. No duplicate records were identified, and transaction
amounts were confirmed to be within reasonable ranges. The datasets were found to be clean and
suitable for feature engineering.


###3) Feature Engineering

####3.1 ) Transactions dataset

In [28]:
#create total_transactions, total_spent, avg_transaction_value, max_transaction_value
transaction_features = transactions.groupby('customer_id').agg(
    total_transactions=('amount', 'count'),
    total_spent=('amount', 'sum'),
    avg_transaction_value=('amount', 'mean'),
    max_transaction_value=('amount', 'max')
).reset_index()

transaction_features.head()

customer_id  total_transactions  total_spent  avg_transaction_value  \
0  CUST000001                   2        88.15              44.075000   
1  CUST000002                   2       245.31             122.655000   
2  CUST000003                   1        18.85              18.850000   
3  CUST000004                   6       126.75              21.125000   
4  CUST000005                   3        67.16              22.386667   

   max_transaction_value  
0                  60.10  
1                 220.31  
2                  18.85  
3                  48.28  
4                  35.91

####3.2 : Chats dataset

In [29]:
#create num_support_chats, avg_chat_sentiment,min_chat_sentiment,min_chat_sentiment
chat_features = chats.groupby('customer_id').agg(
    num_support_chats=('chat_sentiment_score', 'count'),
    avg_chat_sentiment=('chat_sentiment_score', 'mean'),
    min_chat_sentiment=('chat_sentiment_score', 'min'),
    avg_chat_length=('chat_length', 'mean')
).reset_index()

chat_features.head()

customer_id  num_support_chats  avg_chat_sentiment  min_chat_sentiment  \
0  CUST000001                  1               0.546               0.546   
1  CUST000002                  1               0.404               0.404   
2  CUST000003                  1               0.055               0.055   
3  CUST000004                  1               0.088               0.088   
4  CUST000005                  1               0.878               0.878   

   avg_chat_length  
0             55.0  
1             57.0  
2             69.0  
3             74.0  
4             66.0

####3.3 : Reviews dataset (base features)

In [34]:
#create num_reviews,avg_review_rating,avg_review_sentiment,avg_review_length
review_features = reviews.groupby('customer_id').agg(
    num_reviews=('review_rating', 'count'),
    avg_review_rating=('review_rating', 'mean'),
    avg_review_sentiment=('review_sentiment_score', 'mean'),
    avg_review_length=('review_length', 'mean')
).reset_index()

review_features.head()

customer_id  num_reviews  avg_review_rating  avg_review_sentiment  \
0  CUST000001            1                3.0                 0.520   
1  CUST000002            1                3.0                 0.594   
2  CUST000003            1                2.0                 0.055   
3  CUST000004            1                2.0                 0.110   
4  CUST000005            1                4.0                 0.714   

   avg_review_length  
0               34.0  
1               31.0  
2               53.0  
3               53.0  
4               51.0

####3.4 : Reviews Dataset : Burn-risk classification using HuggingFace

In [35]:
!pip -q install transformers
from transformers import pipeline
import pandas as pd

In [36]:
#Load zero-shot classification model
burn_classifier = pipeline(
    "zero-shot-classification",
    model="facebook/bart-large-mnli"
)

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/1.63G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Device set to use cpu


In [37]:
#Define burn-risk labels
labels = ["high churn risk", "low churn risk"]

In [38]:
#test on small sample
sample_reviews = reviews[['customer_id', 'review_text']].head(10).copy()

def classify_burn(text):
    result = burn_classifier(str(text)[:512], labels)
    return result['labels'][0], result['scores'][0]

sample_reviews[['burn_label', 'burn_score']] = sample_reviews['review_text'].apply(
    lambda x: pd.Series(classify_burn(x))
)

sample_reviews[['customer_id', 'burn_label', 'burn_score']]

customer_id       burn_label  burn_score
0  CUST000001   low churn risk    0.989689
1  CUST000002   low churn risk    0.983902
2  CUST000003  high churn risk    0.834812
3  CUST000004  high churn risk    0.965270
4  CUST000005   low churn risk    0.981932
5  CUST000006  high churn risk    0.834812
6  CUST000007   low churn risk    0.972758
7  CUST000008   low churn risk    0.969048
8  CUST000009  high churn risk    0.965270
9  CUST000010   low churn risk    0.930183

In [39]:
#apply to all reviews
burn_results = reviews[['customer_id', 'review_text']].copy()

burn_results[['burn_label', 'burn_score']] = burn_results['review_text'].apply(
    lambda x: pd.Series(classify_burn(x))
)

In [40]:
#convert to numeric features
burn_map = {"high churn risk": 1, "low churn risk": 0}
burn_results['burn_risk_flag'] = burn_results['burn_label'].map(burn_map)

In [41]:
#aggregate to customer lvl
burn_features = burn_results.groupby('customer_id').agg(
    avg_burn_risk=('burn_risk_flag', 'mean'),
    max_burn_risk=('burn_risk_flag', 'max'),
    avg_burn_confidence=('burn_score', 'mean')
).reset_index()

burn_features.head()

customer_id  avg_burn_risk  max_burn_risk  avg_burn_confidence
0  CUST000001            0.0              0             0.989689
1  CUST000002            0.0              0             0.983902
2  CUST000003            1.0              1             0.834812
3  CUST000004            1.0              1             0.965270
4  CUST000005            0.0              0             0.981932